src/View/ViewLayerData.php

```diff
@@ -0,0 +1,254 @@
+<?php
+
+namespace SilverStripe\View;
+
+use BadMethodCallException;
+use InvalidArgumentException;
+use IteratorAggregate;
+use SilverStripe\Core\ClassInfo;
+use SilverStripe\Core\Injector\Injectable;
+use SilverStripe\Model\ModelData;
+use SilverStripe\Model\ModelDataCustomised;
+use SilverStripe\ORM\FieldType\DBClassName;
+use Stringable;
+use Traversable;
+
+class ViewLayerData implements IteratorAggregate, Stringable
+{
+    use Injectable;
+
+    public const TYPE_PROPERTY = 'property';
+
+    public const TYPE_METHOD = 'method';
+
+    public const TYPE_ANY = 'any';
+
+    /**
+     * Special variable names that can be used to get metadata about values
+     */
+    public const META_DATA_NAMES = [
+        // Gets a DBClassName with the class name of $this->data
+        'ClassName',
+        // Returns $this->data
+        'Me',
+    ];
+
+    private object $data;
+
+    public function __construct(mixed $data, mixed $source = null, string $name = '')
+    {
+        if ($data === null) {
+            throw new InvalidArgumentException('$data must not be null');
+        }
+        if ($data instanceof ViewLayerData) {
+            $data = $data->data;
+        } else {
+            $source = $source instanceof ModelData ? $source : null;
+            $data = CastingService::singleton()->cast($data, $source, $name);
+        }
+        $this->data = $data;
+    }
+
+    /**
+     * Needed so we can rewind in SSViewer_Scope::next() after getting itemIteratorTotal without throwing an exception.
+     */
+    public function getIteratorCount(): int
+    {
+        $count = $this->getRawDataValue('count');
+        if (is_numeric($count)) {
+            return $count;
+        }
+        if (is_countable($this->data)) {
+            return count($this->data);
+        }
```

Comment:
This _should_ be redundant to the above, but I've got it here as a matter of completeness

Reply:
Most of these comments are explaining why I've chosen to make a _change_ and are suitable for a PR but not for the code itself. I'm essentially thinking "what questions is Steve likely to ask? I'll answer those before he asks them to speed things up" - these are not "what questions will someone looking at this code _in context_ have in the future?"

Reply:
> GitHub is known to keep Pull Request comments for a long time

When people are looking at code in an IDE, or viewing a file by itself github, or a single commit in github, they won't see the pull request comments.

For many of the comments made, for instance why code was removed, a PR comment does makes sense. Thorugh for lots of others such as https://github.com/silverstripe/silverstripe-framework/pull/11405#discussion_r1811808234 that would make more sense as an inline comments IMO

Inline comments are easier for me the peer reviewer to parse because, due to the large nature of this PR and all the PR comments it came with, things can run very slowly in firefox and I sometimes need to wait several seconds for things to become responsive.